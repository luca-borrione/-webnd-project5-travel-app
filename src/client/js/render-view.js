/* eslint-disable global-require, import/no-dynamic-require */

import {
  formatDateString,
  formatTimestampToLocale,
  getDaysDiff,
  getDaysFromToday,
  getUTCDate,
} from './utils/date-utils';

/**
 * A corner ribbon is rendered on the saved trips.
 * This will show how many days are left from the departure date,
 * or it will indicate that the trip expired.
 */
const renderCornerRibbon = ({ daysFromDeparture }) => {
  const expired = daysFromDeparture < 0;

  let message;
  if (expired) {
    message = 'expired';
  } else if (daysFromDeparture === 0) {
    message = 'today';
  } else if (daysFromDeparture === 1) {
    message = 'tomorrow';
  } else {
    message = `in ${daysFromDeparture} days`;
  }

  return `
    <div class="ribbon ${
      (expired && 'ribbon-top-right') || 'ribbon-top-left'
    }"><span>${message}</span></div>
  `;
};

/**
 * The thumbnail image is rendered together with a caption,
 * which is the name of the city and the flag of its country.
 */
const renderThumbnail = ({
  city,
  flag,
  thumbnail = require('../assets/placeholder-destination-thumb.jpg'),
}) => `
<figure class="card__image">
  <img
    class="card__image-thumbnail"
    src="${thumbnail}"
    alt="view of the destination"
  />
  <figcaption class="card__image-caption">
    <span class="card__image-caption-flag">${flag}</span>
    <span class="card__image-caption-city">${city}</span>
  </figcaption>
</figure>
`;

/**
 * A grid with information about the country of the trip destination.
 */
const renderLocationInfo = ({
  continent,
  subregion,
  country,
  county,
  capital,
  currencies,
  languages,
  timezone,
  offset,
}) =>
  `<article class="card__info">${[
    { label: 'Continent', value: continent },
    { label: 'Subregion', value: subregion },
    { label: 'Country', value: country },
    { label: 'County', value: county },
    { label: 'Capital', value: capital },
    {
      label: 'Currency',
      value: currencies.map(({ code, name }) => `${name} (${code})`).join(', '),
    },
    { label: 'Language', value: languages.join(', ') },
    { label: 'Timezone', value: timezone },
    { label: 'Offset', value: offset },
  ]
    .map(
      ({ label, value }) =>
        value &&
        `
          <div class="card__info-entry">
            <div class="label">${label}:</div>
            <div class="value">${value}</div>
          </div>
        `
    )
    .join('')}</article>`;

/**
 * Helper function to render the weather info:
 * - weather icon and description
 * - temperature in celsius
 * - humidity and wind speed
 */
const renderWeather = ({ description, humidity, icon, temperature, windSpeed }) => `
  <aside class="weather">
    <img
    class="weather__icon"
    src="${require(`../assets/weather-icons/${icon}.png`)}"
    alt="weather icon"
    />
    <div class="weather__temperature">
      <span>${temperature}<sup>&#8451;</sup></span>
    </div>
    <div class="weather__extra-info">
    H ${Math.round(humidity)}% W ${Math.round(windSpeed)}&#x33A7;
    </div>
    <div class="weather__description">${description}</div>
  </aside>`;

/**
 * Helper function to render a box containing a date and the weather info if available
 */
const renderInfoBox = ({ className, dateFormatted, title, weather = '' }) => `
  <article class="infobox ${className}">
    <div class="infobox__title">${title}</div>
    <div class="infobox__date">${dateFormatted}</div>
    ${weather && renderWeather(weather)}
  </article>`;

/**
 * Current locale date and weather at destination
 */
const renderCurrentInfo = ({ timezone, weather }) =>
  renderInfoBox({
    className: 'current',
    title: 'Last Observed',
    dateFormatted: formatTimestampToLocale({
      timestamp: weather.observedTimestamp,
      timezone,
    }),
    weather,
  });

/**
 * Date and weather forecast if available when departing
 */
const renderDepartureInfo = ({ dateString, weather }) =>
  renderInfoBox({
    className: 'departure',
    title: 'Departure',
    dateFormatted: formatDateString(dateString),
    weather,
  });

/**
 * Date and weather forecast if available when returning
 */
const renderReturnInfo = ({ dateString, weather }) =>
  renderInfoBox({
    className: 'return',
    title: 'Return',
    dateFormatted: formatDateString(dateString),
    weather,
  });

const renderDaysDiff = (departureDateString, returnDateString) => `
  <aside class="days-diff">
    <div class="days-diff__title">Length:</div>
    <div class="days-diff__value">${getDaysDiff(departureDateString, returnDateString)} days</div>
  </aside>`;

/**
 * `Save Trip` button rendered on the results trip card
 */
const renderCardButton = () => `
  <button class="card__button" type="button">
    <i class="icon-heart-empty"></i>
    <i class="icon-heart"></i>
    <span>Save Trip</span>
  </button>
  `;

/**
 * `Remove Trip` button rendered on the individual saved trip card
 */
const renderRemoveTripButton = (tripId) => `
  <button class="card__button" type="button" onClick="return Client.removeTrip(${tripId})">
    <i class="icon-heart-empty"></i>
    <i class="icon-heart"></i>
    <span>Remove Trip</span>
  </button>
  `;

/**
 * Rendering the results
 */
export const renderResultsView = ({
  thumbnail,
  locationInfo: { city, flag, ...locationInfo },
  currentInfo,
  departureInfo,
  returnInfo,
}) => `
<main class="card">
  ${renderThumbnail({ thumbnail, city, flag })}
  ${renderLocationInfo(locationInfo)}
  ${renderCurrentInfo(currentInfo)}
  ${renderDepartureInfo(departureInfo)}
  ${renderReturnInfo(returnInfo)}
  ${renderDaysDiff(departureInfo.dateString, returnInfo.dateString)}
  ${renderCardButton()}
</main>`;

/**
 * Saved trips will be rendered sorted by the date by date ascending
 * Example:
 * 21 Dec 2021 - 22 Dec 2021 - 3 Feb 2022
 */
const byDateAscending = (a, b) =>
  getUTCDate(a.departureInfo.dateString) - getUTCDate(b.departureInfo.dateString);

/**
 * Expired trip card will be moved to the end of the list,
 * keeping the ascending order by date
 */
const moveExpiredTripsToTheEnd = (savedTrips) =>
  savedTrips
    .reduce(
      (acc, trip) => {
        if (getDaysFromToday(trip.departureInfo.dateString) < 0) {
          acc[1] = [...acc[1], trip];
        } else {
          acc[0] = [...acc[0], trip];
        }
        return acc;
      },
      [[], []]
    )
    .flat();

/**
 * Helper function to sort the array of the saved list
 */
const sortSavedTrips = (savedTrips) => moveExpiredTripsToTheEnd(savedTrips.sort(byDateAscending));

/**
 * Rendering the whole array of saved trips.
 * Expired trip card will be displayed with a different UI
 */
export const renderSavedTripsView = (savedTrips) =>
  sortSavedTrips(savedTrips).reduce(
    (acc, { id: tripId, thumbnail, locationInfo: { city, flag }, departureInfo, returnInfo }) => {
      const daysFromDeparture = getDaysFromToday(departureInfo.dateString);
      const expired = daysFromDeparture < 0;
      return `
        ${acc}
        <main class="card${(expired && ' expired') || ''}" id="${tripId}">
          ${renderCornerRibbon({ daysFromDeparture })}
          ${renderThumbnail({ thumbnail, city, flag })}
          ${renderDepartureInfo(departureInfo)}
          ${renderReturnInfo(returnInfo)}
          ${renderDaysDiff(departureInfo.dateString, returnInfo.dateString)}
          ${renderRemoveTripButton(tripId)}
        </main>
      `;
    },
    ''
  );

/* eslint-enable global-require, import/no-dynamic-require */
