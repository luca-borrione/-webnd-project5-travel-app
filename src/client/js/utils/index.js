export { getData, postData } from './controller-utils';
export { handleError, handleErrorAndReject } from './error-utils';
