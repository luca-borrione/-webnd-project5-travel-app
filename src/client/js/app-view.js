import { scrollElementIntoView } from './utils/browser-utils';
import { getDaysFromToday } from './utils/date-utils';
import { handleError } from './utils/error-utils';
import {
  getCurrentWeather,
  getGeoName,
  getLocationInfo,
  getThumbnail,
  getWeatherForecast,
  postRemoveTrip,
  postSaveTrip,
  postRestoreTrips,
} from './app-controller';
import { renderResultsView, renderSavedTripsView } from './render-view';

const SAVED_TRIPS_KEY = 'savedTrips';

let $form;
let $results;
let $savedTrips;
let $spinner;
let $main;

const storeElements = () => {
  $form = {
    search: document.forms.search,
    destination: document.forms.search.elements['search-form__destination'],
    departure: document.forms.search.elements['search-form__departure'],
    departureDate: document.forms.search.elements['search-form__departure-date'],
    returnDate: document.forms.search.elements['search-form__return-date'],
    button: document.querySelector('.search-form__submit-button'),
  };

  $results = document.querySelector('.results');
  $savedTrips = document.querySelector('.saved-trips');
  $spinner = document.querySelector('.spinner');
  $main = document.querySelector('.main-area');
};

const toggleSpinner = () => {
  $spinner.classList.toggle('hide');
};

const toggleMain = () => {
  $main.classList.toggle('hide');
};

const toggleSubmitButton = () => {
  if ($form.button.classList.contains('loading')) {
    $form.button.disabled = false;
  } else {
    $form.button.disabled = true;
  }
  $form.button.classList.toggle('loading');
};

const saveTripListener = (trip) => (event) => {
  event.preventDefault();
  event.stopPropagation();
  return postSaveTrip({ trip })
    .then((savedTrips) => {
      window.localStorage.setItem(SAVED_TRIPS_KEY, JSON.stringify(savedTrips));
      $savedTrips.innerHTML = renderSavedTripsView(savedTrips);
    })
    .catch(handleError);
};

export const removeTrip = (tripId) =>
  postRemoveTrip({ tripId })
    .then((savedTrips) => {
      localStorage.setItem(SAVED_TRIPS_KEY, JSON.stringify(savedTrips));
      $savedTrips.innerHTML = renderSavedTripsView(savedTrips);
    })
    .catch(handleError);

const restoreSavedTrips = async () => {
  const localStorageTrips = JSON.parse(window.localStorage.getItem(SAVED_TRIPS_KEY));
  const restoredTrips = await postRestoreTrips({ localStorageTrips });
  window.localStorage.setItem(SAVED_TRIPS_KEY, JSON.stringify(restoredTrips));
  $savedTrips.innerHTML = renderSavedTripsView(restoredTrips);
};

const onSearchFormSubmit = async (event) => {
  event.preventDefault();
  event.stopPropagation();

  toggleSubmitButton();

  try {
    const daysFromDeparture = getDaysFromToday($form.departureDate.value);

    const { latitude, longitude, city, county, country, geonameId } = await getGeoName({
      location: $form.destination.value,
    });

    return Promise.all([
      getLocationInfo({ latitude, longitude }),
      getThumbnail({ city, country }),
      getCurrentWeather({ latitude, longitude }),
      daysFromDeparture < 16
        ? getWeatherForecast({
            latitude,
            longitude,
            departureDate: $form.departureDate.value,
            returnDate: $form.returnDate.value,
          })
        : Promise.resolve(),
    ]).then(
      ([
        { capital, continent, currencies, languages, timezone, offset, flag, subregion },
        thumbnail,
        { dateString: currentDateString, ...currentWeather },
        forecast,
      ]) => {
        const tripCard = {
          id: new Date($form.departureDate.value).getTime() + geonameId,
          thumbnail,
          locationInfo: {
            capital,
            city,
            continent,
            country,
            county,
            currencies,
            flag,
            languages,
            latitude,
            longitude,
            offset,
            subregion,
            timezone,
          },
          currentInfo: {
            dateString: currentDateString,
            timezone,
            weather: currentWeather,
          },
          departureInfo: {
            dateString: $form.departureDate.value,
            weather: forecast?.departureWeather,
          },
          returnInfo: {
            dateString: $form.returnDate.value,
            weather: forecast?.returnWeather,
          },
        };

        $results.innerHTML = renderResultsView(tripCard);

        $results
          .querySelector('.card__button')
          .addEventListener('click', saveTripListener(tripCard));

        toggleSubmitButton();
        scrollElementIntoView($results);
      }
    );
  } catch (e) {
    return handleError(e);
  }
};

const initEventListeners = () => {
  $form.search.addEventListener('submit', onSearchFormSubmit);
};

const init = async () => {
  try {
    storeElements();
    initEventListeners();
    await restoreSavedTrips();
    toggleSpinner();
    toggleMain();
  } catch (error) {
    handleError(error);
  }
};

window.document.addEventListener('DOMContentLoaded', init);
