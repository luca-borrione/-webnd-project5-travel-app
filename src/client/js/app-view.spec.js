import {
  getCurrentWeather,
  getGeoName,
  getLocationInfo,
  getThumbnail,
  getWeatherForecast,
  postRemoveTrip,
  postSaveTrip,
  postRestoreTrips,
} from './app-controller';
import { scrollElementIntoView } from './utils/browser-utils';
import { getDaysFromToday } from './utils/date-utils';
import { handleError } from './utils/error-utils';
import { renderResultsView, renderSavedTripsView } from './render-view';
import { removeTrip } from './app-view';

jest.mock('./../assets/placeholder-destination-thumb.jpg', () => 'mock-placeholder-image');

jest.mock('./utils/date-utils', () => ({
  getDaysFromToday: jest.fn().mockReturnValue(1),
}));

jest.mock('./utils/error-utils');

jest.mock('./utils/browser-utils');

jest.mock('./app-controller', () => ({
  getGeoName: jest.fn().mockResolvedValue({
    county: 'mock-county',
    city: 'mock-city',
    latitude: 'mock-latitude',
    longitude: 'mock-longitude',
    country: 'mock-country',
    geonameId: 1234,
  }),
  getLocationInfo: jest.fn().mockResolvedValue({
    capital: 'mock-capital',
    continent: 'mock-continent',
    currencies: [
      { name: 'mock-currency1', code: 'cur1' },
      { name: 'mock-currency2', code: 'cur2' },
    ],
    languages: ['mock-language1', 'mock-language2'],
    timezone: 'mock-timezone',
    offset: 'mock-offset',
    flag: 'mock-flag',
    subregion: 'mock-subregion',
  }),
  getThumbnail: jest.fn().mockResolvedValue('mock-image-url'),
  getCurrentWeather: jest.fn().mockResolvedValue({
    dateString: 'mock-date-string',
    description: 'mock-description',
    humidity: 'mock-humidity',
    icon: 'mock-icon',
    temperature: 'mock-temperature',
    timezone: 'mock-timezone',
    windSpeed: 'mock-wind-speed',
  }),
  getWeatherForecast: jest.fn().mockResolvedValue({
    departure: {
      description: 'mock-description',
      humidity: 'mock-humidity',
      icon: 'mock-icon',
      windSpeed: 'mock-wind-speed',
    },
    return: {
      description: 'mock-description',
      humidity: 'mock-humidity',
      icon: 'mock-icon',
      windSpeed: 'mock-wind-speed',
    },
  }),
  postSaveTrip: jest.fn().mockResolvedValue([{ saved: 'trip1' }, { saved: 'trip2' }]),
  postRemoveTrip: jest.fn().mockResolvedValue([{ saved: 'trip2' }]),
  postRestoreTrips: jest.fn().mockResolvedValue([]),
}));

jest.mock('./render-view', () => ({
  renderResultsView: jest.fn().mockReturnValue(`
    <main class="card">
      mock-results-view
      <button class="card__button" type="button"></button>
    </main>
  `),
  renderSavedTripsView: jest.fn().mockReturnValue(`
    <main class="card">
      mock-saved-trips-view
    </main>
  `),
}));

const localStorageProto = Object.getPrototypeOf(window.localStorage);
jest.spyOn(localStorageProto, 'getItem').mockReturnValue('[{ "retrievedFrom": "storage" }]');
jest.spyOn(localStorageProto, 'setItem').mockImplementation();

// eslint-disable-next-line no-promise-executor-return
const flushPromises = () => new Promise((resolve) => setImmediate(resolve));

describe('app-view', () => {
  const SAVED_TRIPS_KEY = 'savedTrips';

  let $form;
  let $main;
  let $results;
  let $savedTrips;
  let $spinner;

  const storeElements = () => {
    $form = {
      search: document.forms.search,
      destination: document.forms.search.elements['search-form__destination'],
      departure: document.forms.search.elements['search-form__departure'],
      departureDate: document.forms.search.elements['search-form__departure-date'],
      returnDate: document.forms.search.elements['search-form__return-date'],
      button: document.querySelector('.search-form__submit-button'),
    };

    $main = document.querySelector('main');
    $results = document.querySelector('.results');
    $savedTrips = document.querySelector('.saved-trips');
    $spinner = document.querySelector('.spinner');
  };

  beforeEach(() => {
    window.document.body.innerHTML = `
      <form name="search">
        <input id="search-form__destination" type="text" />
        <input id="search-form__departure" type="text" />
        <input id="search-form__departure-date" type="date" />
        <input id="search-form__return-date" type="date" />
        <button type="submit" class="search-form__submit-button" />
      </form>
      <aside class="spinner" />
      <main class="main-area hide">
        <section class="results" />
        <section class="saved-trips" />
      </main>
    `;

    storeElements();
  });

  describe('launching the app', () => {
    it('should hide the main area of the app and showing a spinner, while restoring the saved trips', async () => {
      expect($main.classList.contains('hide')).toBe(true);
      expect($spinner.classList.contains('hide')).toBe(false);
      window.document.dispatchEvent(
        new Event('DOMContentLoaded', {
          bubbles: true,
          cancelable: true,
        })
      );
      await flushPromises();
      expect($main.classList.contains('hide')).toBe(false);
      expect($spinner.classList.contains('hide')).toBe(true);
    });

    it('should try to retrieve the saved trips from the local storage', async () => {
      expect(window.localStorage.getItem).not.toHaveBeenCalled();
      window.document.dispatchEvent(
        new Event('DOMContentLoaded', {
          bubbles: true,
          cancelable: true,
        })
      );
      await flushPromises();
      expect(window.localStorage.getItem).toHaveBeenCalledTimes(1);
      expect(window.localStorage.getItem).toHaveBeenCalledWith(SAVED_TRIPS_KEY);
    });

    it('should correctly call postRestoreTrips', async () => {
      expect(postRestoreTrips).not.toHaveBeenCalled();
      window.document.dispatchEvent(
        new Event('DOMContentLoaded', {
          bubbles: true,
          cancelable: true,
        })
      );
      await flushPromises();
      expect(postRestoreTrips).toHaveBeenCalledTimes(1);
      expect(postRestoreTrips).toHaveBeenCalledWith({
        localStorageTrips: [{ retrievedFrom: 'storage' }],
      });
    });

    it('should save to the local storage the restored and updated trips received from the server', async () => {
      postRestoreTrips.mockResolvedValueOnce([{ restoredFrom: 'server' }]);
      expect(window.localStorage.setItem).not.toHaveBeenCalled();
      window.document.dispatchEvent(
        new Event('DOMContentLoaded', {
          bubbles: true,
          cancelable: true,
        })
      );
      await flushPromises();
      expect(window.localStorage.setItem).toHaveBeenCalledTimes(1);
      expect(window.localStorage.setItem).toHaveBeenCalledWith(
        SAVED_TRIPS_KEY,
        '[{"restoredFrom":"server"}]'
      );
    });

    it('should correctly call renderSavedTripsView', async () => {
      postRestoreTrips.mockResolvedValueOnce([{ restoredFrom: 'server' }]);
      expect(renderSavedTripsView).not.toHaveBeenCalled();
      window.document.dispatchEvent(
        new Event('DOMContentLoaded', {
          bubbles: true,
          cancelable: true,
        })
      );
      await flushPromises();
      expect(renderSavedTripsView).toHaveBeenCalledTimes(1);
      expect(renderSavedTripsView).toHaveBeenCalledWith([{ restoredFrom: 'server' }]);
    });

    it('should correctly update the saved trips view', async () => {
      expect($savedTrips).toMatchSnapshot();
      window.document.dispatchEvent(
        new Event('DOMContentLoaded', {
          bubbles: true,
          cancelable: true,
        })
      );
      await flushPromises();
      expect($savedTrips).toMatchSnapshot();
    });

    it('should handle an error nicely, if postRestoreTrips rejects', async () => {
      const expectedError = new Error('mock-expected-error');
      postRestoreTrips.mockRejectedValueOnce(expectedError);
      expect(handleError).not.toHaveBeenCalled();
      window.document.dispatchEvent(
        new Event('DOMContentLoaded', {
          bubbles: true,
          cancelable: true,
        })
      );
      await flushPromises();
      expect(handleError).toHaveBeenCalledTimes(1);
      expect(handleError).toHaveBeenCalledWith(expectedError);
    });
  });

  describe('after having launched the app', () => {
    beforeEach(async () => {
      window.document.dispatchEvent(
        new Event('DOMContentLoaded', {
          bubbles: true,
          cancelable: true,
        })
      );

      await flushPromises();
      jest.clearAllMocks();
      $savedTrips.innerHTML = '';

      $form.destination.value = 'mock-destination';
      $form.departure.value = 'mock-departure';
      $form.departureDate.value = '2021-11-10';
      $form.returnDate.value = '2021-11-17';
    });

    describe('submitting a new search', () => {
      it('should put the form submit button into a loading state', () => {
        expect($form.button.classList.contains('loading')).toBe(false);
        expect($form.button.disabled).toBe(false);
        $form.search.submit();
        expect($form.button.classList.contains('loading')).toBe(true);
        expect($form.button.disabled).toBe(true);
      });

      it('should correctly call getGeoName', () => {
        expect(getGeoName).not.toHaveBeenCalled();
        $form.search.submit();
        expect(getGeoName).toHaveBeenCalledTimes(1);
        expect(getGeoName).toHaveBeenCalledWith({ location: 'mock-destination' });
      });

      it('should correctly call getLocationInfo', async () => {
        expect(getLocationInfo).not.toHaveBeenCalled();
        $form.search.submit();
        await flushPromises();
        expect(getLocationInfo).toHaveBeenCalledTimes(1);
        expect(getLocationInfo).toHaveBeenCalledWith({
          latitude: 'mock-latitude',
          longitude: 'mock-longitude',
        });
      });

      it('should correctly call getThumbnail', async () => {
        expect(getThumbnail).not.toHaveBeenCalled();
        $form.search.submit();
        await flushPromises();
        expect(getThumbnail).toHaveBeenCalledTimes(1);
        expect(getThumbnail).toHaveBeenCalledWith({
          country: 'mock-country',
          city: 'mock-city',
        });
      });

      it('should correctly call getCurrentWeather', async () => {
        expect(getCurrentWeather).not.toHaveBeenCalled();
        $form.search.submit();
        await flushPromises();
        expect(getCurrentWeather).toHaveBeenCalledTimes(1);
        expect(getCurrentWeather).toHaveBeenCalledWith({
          latitude: 'mock-latitude',
          longitude: 'mock-longitude',
        });
      });

      it('should correctly call getWeatherForecast when the departure date is within 16 days from today', async () => {
        expect(getWeatherForecast).not.toHaveBeenCalled();
        $form.search.submit();
        await flushPromises();
        expect(getWeatherForecast).toHaveBeenCalledTimes(1);
        expect(getWeatherForecast).toHaveBeenCalledWith({
          latitude: 'mock-latitude',
          longitude: 'mock-longitude',
          departureDate: '2021-11-10',
          returnDate: '2021-11-17',
        });
      });

      it('should not call getWeatherForecast when the departure date is at least 16 days from today', async () => {
        getDaysFromToday.mockReturnValueOnce(16);
        expect(getWeatherForecast).not.toHaveBeenCalled();
        $form.search.submit();
        await flushPromises();
        expect(getWeatherForecast).not.toHaveBeenCalled();
      });
    });

    describe('receiving the data of the search', () => {
      it.each`
        fnName                  | fn
        ${'getGeoName'}         | ${getGeoName}
        ${'getLocationInfo'}    | ${getLocationInfo}
        ${'getThumbnail'}       | ${getThumbnail}
        ${'getCurrentWeather'}  | ${getCurrentWeather}
        ${'getWeatherForecast'} | ${getWeatherForecast}
      `('should handle an error nicely, if $fnName rejects', async ({ fn }) => {
        const expectedError = new Error('mock-expected-error');
        fn.mockRejectedValueOnce(expectedError);
        expect(handleError).not.toHaveBeenCalled();
        $form.search.submit();
        await flushPromises();
        expect(handleError).toHaveBeenCalledTimes(1);
        expect(handleError).toHaveBeenCalledWith(expectedError);
      });

      it('should correctly call renderResultsView', async () => {
        expect(renderResultsView).not.toHaveBeenCalled();
        $form.search.submit();
        await flushPromises();
        expect(renderResultsView).toHaveBeenCalledTimes(1);
        expect(renderResultsView.mock.calls[0][0]).toMatchSnapshot();
      });

      it('should correctly update the results view', async () => {
        expect($results).toMatchSnapshot();
        $form.search.submit();
        await flushPromises();
        expect($results).toMatchSnapshot();
      });

      it('should restore the form submit button to its initial state', async () => {
        expect($form.button.classList.contains('loading')).toBe(false);
        expect($form.button.disabled).toBe(false);
        $form.search.submit();
        expect($form.button.classList.contains('loading')).toBe(true);
        expect($form.button.disabled).toBe(true);
        await flushPromises();
        expect($form.button.classList.contains('loading')).toBe(false);
        expect($form.button.disabled).toBe(false);
      });

      it('schould try to scroll the results into view', async () => {
        expect(scrollElementIntoView).not.toHaveBeenCalled();
        $form.search.submit();
        await flushPromises();
        expect(scrollElementIntoView).toHaveBeenCalledTimes(1);
        expect(scrollElementIntoView).toHaveBeenCalledWith($results);
      });
    });

    describe('clicking on the save trip button of the search results', () => {
      let $saveTripButton;

      beforeEach(async () => {
        $form.search.submit();
        await flushPromises();
        $saveTripButton = $results.querySelector('.card__button');
      });

      it('should correctly call postSaveTrip', async () => {
        expect(postSaveTrip).not.toHaveBeenCalled();
        $saveTripButton.click();
        await flushPromises();
        expect(postSaveTrip).toHaveBeenCalledTimes(1);
        expect(postSaveTrip.mock.calls[0][0]).toMatchSnapshot();
      });

      it('should store in the local storage the trips saved on the server', async () => {
        expect(window.localStorage.setItem).not.toHaveBeenCalled();
        $saveTripButton.click();
        await flushPromises();
        expect(window.localStorage.setItem).toHaveBeenCalledTimes(1);
        expect(window.localStorage.setItem).toHaveBeenCalledWith(
          SAVED_TRIPS_KEY,
          '[{"saved":"trip1"},{"saved":"trip2"}]'
        );
      });

      it('should correctly call renderSavedTripsView', async () => {
        expect(renderSavedTripsView).not.toHaveBeenCalled();
        $saveTripButton.click();
        await flushPromises();
        expect(renderSavedTripsView).toHaveBeenCalledTimes(1);
        expect(renderSavedTripsView).toHaveBeenCalledWith([{ saved: 'trip1' }, { saved: 'trip2' }]);
      });

      it('should correctly update the saved trips view', async () => {
        expect($savedTrips).toMatchSnapshot();
        $saveTripButton.click();
        await flushPromises();
        expect($savedTrips).toMatchSnapshot();
      });

      it('should handle an error nicely, if postSaveTrip rejects', async () => {
        const expectedError = new Error('mock-expected-error');
        postSaveTrip.mockRejectedValueOnce(expectedError);
        expect(handleError).not.toHaveBeenCalled();
        $saveTripButton.click();
        await flushPromises();
        expect(handleError).toHaveBeenCalledTimes(1);
        expect(handleError).toHaveBeenCalledWith(expectedError);
      });
    });

    describe('calling removeTrip', () => {
      it('should correcly call postRemoveTrip', async () => {
        expect(postRemoveTrip).not.toHaveBeenCalled();
        removeTrip('mock-trip-id');
        await flushPromises();
        expect(postRemoveTrip).toHaveBeenCalledTimes(1);
        expect(postRemoveTrip).toHaveBeenCalledWith({ tripId: 'mock-trip-id' });
      });

      it('should store in the local storage the remaining trips saved on the server', async () => {
        expect(window.localStorage.setItem).not.toHaveBeenCalled();
        removeTrip('mock-trip-id');
        await flushPromises();
        expect(window.localStorage.setItem).toHaveBeenCalledTimes(1);
        expect(window.localStorage.setItem).toHaveBeenCalledWith(
          SAVED_TRIPS_KEY,
          '[{"saved":"trip2"}]'
        );
      });

      it('should correctly call renderSavedTripsView', async () => {
        expect(renderSavedTripsView).not.toHaveBeenCalled();
        removeTrip('mock-trip-id');
        await flushPromises();
        expect(renderSavedTripsView).toHaveBeenCalledTimes(1);
        expect(renderSavedTripsView).toHaveBeenCalledWith([{ saved: 'trip2' }]);
      });

      it('should correctly update the saved trips view', async () => {
        expect($savedTrips).toMatchSnapshot();
        removeTrip('mock-trip-id');
        await flushPromises();
        expect($savedTrips).toMatchSnapshot();
      });

      it('should handle an error nicely, if postRemoveTrip rejects', async () => {
        const expectedError = new Error('mock-expected-error');
        postRemoveTrip.mockRejectedValueOnce(expectedError);
        expect(handleError).not.toHaveBeenCalled();
        removeTrip('mock-trip-id');
        await flushPromises();
        expect(handleError).toHaveBeenCalledTimes(1);
        expect(handleError).toHaveBeenCalledWith(expectedError);
      });
    });
  });
});
