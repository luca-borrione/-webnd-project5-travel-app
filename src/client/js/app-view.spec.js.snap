// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app-view after having launched the app calling removeTrip should correctly update the saved trips view 1`] = `
<section
  class="saved-trips"
/>
`;

exports[`app-view after having launched the app calling removeTrip should correctly update the saved trips view 2`] = `
<section
  class="saved-trips"
>
  
    
  <main
    class="card"
  >
    
      mock-saved-trips-view
    
  </main>
  
  
</section>
`;

exports[`app-view after having launched the app clicking on the save trip button of the search results should correctly call postSaveTrip 1`] = `
Object {
  "trip": Object {
    "currentInfo": Object {
      "dateString": "mock-date-string",
      "timezone": "mock-timezone",
      "weather": Object {
        "description": "mock-description",
        "humidity": "mock-humidity",
        "icon": "mock-icon",
        "temperature": "mock-temperature",
        "timezone": "mock-timezone",
        "windSpeed": "mock-wind-speed",
      },
    },
    "departureInfo": Object {
      "dateString": "2021-11-10",
      "weather": Object {
        "description": "mock-description",
        "humidity": "mock-humidity",
        "icon": "mock-icon",
        "windSpeed": "mock-wind-speed",
      },
    },
    "id": 1636502401234,
    "locationInfo": Object {
      "capital": "mock-capital",
      "city": "mock-city",
      "continent": "mock-continent",
      "country": "mock-country",
      "county": "mock-county",
      "currencies": Array [
        Object {
          "code": "cur1",
          "name": "mock-currency1",
        },
        Object {
          "code": "cur2",
          "name": "mock-currency2",
        },
      ],
      "flag": "mock-flag",
      "languages": Array [
        "mock-language1",
        "mock-language2",
      ],
      "latitude": "mock-latitude",
      "longitude": "mock-longitude",
      "offset": "mock-offset",
      "subregion": "mock-subregion",
      "timezone": "mock-timezone",
    },
    "returnInfo": Object {
      "dateString": "2021-11-17",
      "weather": Object {
        "description": "mock-description",
        "humidity": "mock-humidity",
        "icon": "mock-icon",
        "windSpeed": "mock-wind-speed",
      },
    },
    "thumbnail": "mock-image-url",
  },
}
`;

exports[`app-view after having launched the app clicking on the save trip button of the search results should correctly update the saved trips view 1`] = `
<section
  class="saved-trips"
/>
`;

exports[`app-view after having launched the app clicking on the save trip button of the search results should correctly update the saved trips view 2`] = `
<section
  class="saved-trips"
>
  
    
  <main
    class="card"
  >
    
      mock-saved-trips-view
    
  </main>
  
  
</section>
`;

exports[`app-view after having launched the app receiving the data of the search should correctly call renderResultsView 1`] = `
Object {
  "currentInfo": Object {
    "dateString": "mock-date-string",
    "timezone": "mock-timezone",
    "weather": Object {
      "description": "mock-description",
      "humidity": "mock-humidity",
      "icon": "mock-icon",
      "temperature": "mock-temperature",
      "timezone": "mock-timezone",
      "windSpeed": "mock-wind-speed",
    },
  },
  "departureInfo": Object {
    "dateString": "2021-11-10",
    "weather": Object {
      "description": "mock-description",
      "humidity": "mock-humidity",
      "icon": "mock-icon",
      "windSpeed": "mock-wind-speed",
    },
  },
  "id": 1636502401234,
  "locationInfo": Object {
    "capital": "mock-capital",
    "city": "mock-city",
    "continent": "mock-continent",
    "country": "mock-country",
    "county": "mock-county",
    "currencies": Array [
      Object {
        "code": "cur1",
        "name": "mock-currency1",
      },
      Object {
        "code": "cur2",
        "name": "mock-currency2",
      },
    ],
    "flag": "mock-flag",
    "languages": Array [
      "mock-language1",
      "mock-language2",
    ],
    "latitude": "mock-latitude",
    "longitude": "mock-longitude",
    "offset": "mock-offset",
    "subregion": "mock-subregion",
    "timezone": "mock-timezone",
  },
  "returnInfo": Object {
    "dateString": "2021-11-17",
    "weather": Object {
      "description": "mock-description",
      "humidity": "mock-humidity",
      "icon": "mock-icon",
      "windSpeed": "mock-wind-speed",
    },
  },
  "thumbnail": "mock-image-url",
}
`;

exports[`app-view after having launched the app receiving the data of the search should correctly update the results view 1`] = `
<section
  class="results"
>
  
        
  <section
    class="saved-trips"
  />
</section>
`;

exports[`app-view after having launched the app receiving the data of the search should correctly update the results view 2`] = `
<section
  class="results"
>
  
    
  <main
    class="card"
  >
    
      mock-results-view
      
    <button
      class="card__button"
      type="button"
    />
    
    
  </main>
  
  
</section>
`;

exports[`app-view launching the app should correctly update the saved trips view 1`] = `
<section
  class="saved-trips"
>
  
      
</section>
`;

exports[`app-view launching the app should correctly update the saved trips view 2`] = `
<section
  class="saved-trips"
>
  
    
  <main
    class="card"
  >
    
      mock-saved-trips-view
    
  </main>
  
  
</section>
`;
